source/extensions/access_loggers/grpc/grpc_access_log_impl.cc

```diff
@@ -0,0 +1,113 @@
+                                           std::chrono::milliseconds buffer_flush_interval_msec,
+                                           uint64_t buffer_size_bytes,
+                                           Event::Dispatcher& dispatcher,
+                                           const LocalInfo::LocalInfo& local_info)
+    : client_(std::move(client)), log_name_(log_name),
+      buffer_flush_interval_msec_(buffer_flush_interval_msec),
+      flush_timer_(dispatcher.createTimer([this]() {
+        flush();
+        flush_timer_->enableTimer(buffer_flush_interval_msec_);
+      })),
+      buffer_size_bytes_(buffer_size_bytes), local_info_(local_info) {
+  flush_timer_->enableTimer(buffer_flush_interval_msec_);
+}
+
+void GrpcAccessLoggerImpl::log(envoy::data::accesslog::v2::HTTPAccessLogEntry&& entry) {
+  approximate_message_size_bytes_ += entry.ByteSizeLong();
+  message_.mutable_http_logs()->add_log_entry()->Swap(&entry);
+  if (approximate_message_size_bytes_ >= buffer_size_bytes_) {
+    flush();
+  }
+}
+
+void GrpcAccessLoggerImpl::flush() {
+  if (!message_.has_http_logs()) {
+    // Nothing to flush.
+    return;
+  }
+
+  if (stream_ == absl::nullopt) {
+    stream_.emplace(*this);
+  }
+
+  if (stream_->stream_ == nullptr) {
+    stream_->stream_ =
+        client_->start(*Protobuf::DescriptorPool::generated_pool()->FindMethodByName(
+                           "envoy.service.accesslog.v2.AccessLogService.StreamAccessLogs"),
+                       *stream_);
+
+    auto* identifier = message_.mutable_identifier();
+    *identifier->mutable_node() = local_info_.node();
+    identifier->set_log_name(log_name_);
+  }
+
+  if (stream_->stream_ != nullptr) {
+    stream_->stream_->sendMessage(message_, false);
+  } else {
+    // Clear out the stream data due to stream creation failure.
+    stream_.reset();
+  }
+
+  // Clear the message regardless of the success.
+  approximate_message_size_bytes_ = 0;
+  message_.Clear();
+}
+
+GrpcAccessLoggerCacheImpl::GrpcAccessLoggerCacheImpl(Grpc::AsyncClientManager& async_client_manager,
+                                                     Stats::Scope& scope,
+                                                     ThreadLocal::SlotAllocator& tls,
+                                                     const LocalInfo::LocalInfo& local_info)
+    : async_client_manager_(async_client_manager), scope_(scope), tls_slot_(tls.allocateSlot()),
+      local_info_(local_info) {
+  tls_slot_->set(
+      [](Event::Dispatcher& dispatcher) { return std::make_shared<ThreadLocalCache>(dispatcher); });
+}
+
+GrpcAccessLoggerSharedPtr GrpcAccessLoggerCacheImpl::getOrCreateLogger(
+    const envoy::config::accesslog::v2::CommonGrpcAccessLogConfig& config) {
+  // TODO(euroelessar): Consider cleaning up loggers.
+  auto& cache = tls_slot_->getTyped<ThreadLocalCache>();
+  const std::size_t cache_key = MessageUtil::hash(config);
```

Comment:
With the separation in place key should include a notion if it's http or tcp logger. Otherwise Envoy can not ensure API promise anymore (that each logging stream is always either http or tcp).

Reply:
Yes, thanks for the reminder, added a TODO here and will add the types when I implement TCP logger.